### PR TITLE
Improve upper bound estimate in broadcast when masks are present

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes scipp/esssans#89.

### CODE REVIEW

File: src/ess/reduce/uncertainty.py [hunk 1]
```diff
@@ -36,50 +38,99 @@
...
+    Returns
+    -------
+    :
+        The data with the variances scaled by the volume of the new subspace.
+    """
+    if _no_variance_broadcast(data, prototype.sizes):
         return data
-    size = 1
+    sizes = {dim: size for dim, size in prototype.sizes.items() if dim not in data.dims}
+    sizes = prototype.sizes
```
Comment (nvaytet) [inline src/ess/reduce/uncertainty.py:79]: Is there a typo here? redefining `sizes`?

File: src/ess/reduce/uncertainty.py [hunk 2]
```diff
@@ -36,50 +38,99 @@
...
+        if dim not in data.dims and dim not in mask.dims:
             size *= dim_size
     data = data.copy()
     data.variances *= size
-    return data.broadcast(sizes={**sizes, **data.sizes}).copy()
+    sizes = {**sizes, **data.sizes}
+    data = data.broadcast(sizes=sizes).copy()
+    if mask is not None:
+        # The masked values are not counted in the variance, so we set them to infinity.
+        data.variances[mask.broadcast(sizes=sizes).values] = np.inf
```
Comment (nvaytet) [inline src/ess/reduce/uncertainty.py:97]: What's the reasoning behind `np.inf`? I know the masked data should not be used, is `np.nan` a better choice? Inf maybe be annoying to handle in plots, fx.

Comment (SimonHeybrock) [reply src/ess/reduce/uncertainty.py:97]: I went by my gut feeling "I am making some error bars smaller, so the others should be larger, to balance it out". Wouldn't scipp/plopp#3292 have solved the issue with plotting you mention?

That being said, if you think NaN actually makes more sense then I can change it.

File: src/ess/reduce/uncertainty.py (hunk 1 above)
Comment (SimonHeybrock) [reply src/ess/reduce/uncertainty.py:79]: Oops, first line was left over from refactor.

File: src/ess/reduce/uncertainty.py (hunk 2 above)
Comment (nvaytet) [reply src/ess/reduce/uncertainty.py:96]: Actually, it makes no difference for the plotting, so you can leave it if you like. I though you would get vertical lines spanning the entire height of the figure for the error bars, but in fact nothing is plotted (both my MPL and plotly).
Screenshot at 2024-06-21 10-51-06

I don't know what is a better indicator that the data is invalid there: if it's NaN, does it appear more 'intentional' than inf which could be a result from an operation gone wrong?
In any case, I think it's ok either way.

**Review by nvaytet — APPROVED** [review]

**Pull request merged**